pkg/controller/instance/plan_execution_engine.go

```diff
@@ -1,332 +1,228 @@
-	if isInProgress(state.Status) {
-		state.Status = v1alpha1.ExecutionInProgress
-
-		// check if step is already healthy
-		allHealthy := true
-		for _, r := range resources {
-			if step.Delete {
-				// delete
-				log.Printf("PlanExecution: Step %s will delete object %v", step.Name, r)
-				err := c.Delete(context.TODO(), r, client.PropagationPolicy(metav1.DeletePropagationForeground))
-				if !apierrors.IsNotFound(err) && err != nil {
-					return err
+				// - 3.c build task context -
+				ctx := engtask.Context{
+					Client:     c,
+					Enhancer:   enh,
+					Meta:       exm,
+					Templates:  pl.templates,
+					Parameters: pl.params,
 				}
-			} else {
-				// create or update
-				log.Printf("Going to create/update %v", r)
-				existingResource := r.DeepCopyObject()
-				key, _ := client.ObjectKeyFromObject(r)
-				err := c.Get(context.TODO(), key, existingResource)
-				if apierrors.IsNotFound(err) {
-					// create
-					err = c.Create(context.TODO(), r)
-					if err != nil {
-						log.Printf("PlanExecution: error when creating resource in step %v: %v", step.Name, err)
-						return err
-					}
-				} else if err != nil {
-					// other than not found error - raise it
-					return err
-				} else {
-					// update
-					err := patchExistingObject(r, existingResource, c)
-					if err != nil {
-						return err
+
+				// --- 4. Execute the engine task ---
+				done, err := task.Run(ctx)
+
+				// a fatal error is propagated through the plan/phase/step statuses and the plan execution will be
+				// stopped in the spirit of "fail-loud-and-proud".
+				switch {
+				case errors.Is(err, engtask.ErrFatalExecution):
+					log.Printf("PlanExecution: fatal error during task %s execution for operator version %s: %v", exm.TaskName, exm.OperatorVersionName, err)
+					phaseStatus.Status = v1alpha1.ExecutionFatalError
+					stepStatus.Status = v1alpha1.ExecutionFatalError
+					planStatus.Status = v1alpha1.ExecutionFatalError
+					return planStatus, ExecutionError{
+						Err:       fmt.Errorf("fatal task %s execution error  for operator version %s: %w", tn, em.OperatorVersionName, err),
+						Fatal:     true,
+						EventName: &fatalTaskExecutionErrorEventName,
 					}
+				case err != nil:
+					log.Printf("PlanExecution: error during task %s execution for operator version %s: %v", exm.TaskName, exm.OperatorVersionName, err)
+					stepStatus.Status = v1alpha1.ErrorStatus
+				case done:
+					tasksLeft = tasksLeft - 1
 				}
+			}
 
-				err = health.IsHealthy(c, existingResource)
-				if err != nil {
-					allHealthy = false
-					log.Printf("PlanExecution: Obj is NOT healthy: %s", prettyPrint(key))
+			// --- 5. Check if all TASKs are finished ---
+			// if some TASKs aren't ready yet and STEPs strategy is serial we can not proceed
+			// otherwise, if STEPs strategy is parallel or all TASKs are finished, we can go to the next STEP
+			if tasksLeft > 0 {
```

Comment:
This is kind of personal taste but it would definitely make it easier for me as reader if we were explicit here:
if tasksLeft > 0 && ph.Strategy == v1alpha1.Serial
if tasksLeft > 0 && ph.Strategy == v1alpha1.Parallel (just continue I guess?)
else ...

No strong feelings around this but I spent some time wondering where does the inner else branch end up :)

Reply:
wait on a different note - how does phase strategy affect tasks? I am still struggling with the levels but phase strategy only affect execution of steps, right? So should it even be in here?

Reply:
>  how does phase strategy affect tasks?

Lol, this gets me every time. TL;DR:
- `phase.Strategy` is for *the steps inside the phase*. Totally logical, I know :D 
- failed task `->` means failed step, hence, `phase.Strategy` has to be consulted

Reply:
> This is kind of personal taste but it would definitely make it easier for me as reader if we were explicit here:

We don't need to handle the parallel case - that's where we just continue. And if you leave that out, what's left is:
```
			if tasksLeft > 0 && ph.Strategy == v1alpha1.Serial {
				log.Printf("PlanExecution: some tasks of the %s.%s, operator version %s are not ready", ph.Name, st.Name, em.OperatorVersionName)
				break
			} else {
				stepStatus.Status = v1alpha1.ExecutionComplete
				stepsLeft = stepsLeft - 1
			}
```

So, the only difference is that `tasksLeft > 0 && ph.Strategy == v1alpha1.Serial ` is in one line. I can do it! 👍